js/flexbuffers.js

```diff
@@ -0,0 +1,1015 @@
+};
+
+flexbuffers.ValueType.isIndirectNumber = (value) => {
+  return value >= flexbuffers.ValueType.INDIRECT_INT
+    && value <= flexbuffers.ValueType.INDIRECT_FLOAT;
+};
+
+flexbuffers.ValueType.isTypedVectorElement = (value) => {
+  return value === flexbuffers.ValueType.BOOL
+    || (value >= flexbuffers.ValueType.INT
+      && value <= flexbuffers.ValueType.STRING);
+};
+
+flexbuffers.ValueType.isTypedVector = (value) => {
+  return value === flexbuffers.ValueType.VECTOR_BOOL
+    || (value >= flexbuffers.ValueType.VECTOR_INT
+      && value <= flexbuffers.ValueType.VECTOR_STRING_DEPRECATED);
+};
+
+flexbuffers.ValueType.isFixedTypedVector = (value) => {
+  return value >= flexbuffers.ValueType.VECTOR_INT2
+    && value <= flexbuffers.ValueType.VECTOR_FLOAT4;
+};
+
+flexbuffers.ValueType.isAVector = (value) => {
+  return flexbuffers.ValueType.isTypedVector(value)
+    || flexbuffers.ValueType.isFixedTypedVector(value)
+    || value === flexbuffers.ValueType.VECTOR;
+};
+
+flexbuffers.ValueType.toTypedVector = (valueType, length) => {
+  if (length === 0) return valueType - flexbuffers.ValueType.INT + flexbuffers.ValueType.VECTOR_INT;
+  if (length === 2) return valueType - flexbuffers.ValueType.INT + flexbuffers.ValueType.VECTOR_INT2;
+  if (length === 3) return valueType - flexbuffers.ValueType.INT + flexbuffers.ValueType.VECTOR_INT3;
+  if (length === 4) return valueType - flexbuffers.ValueType.INT + flexbuffers.ValueType.VECTOR_INT4;
+  throw "Unexpected length " + length;
+};
+
+flexbuffers.ValueType.typedVectorElementType = (valueType) => {
+  return valueType - flexbuffers.ValueType.VECTOR_INT + flexbuffers.ValueType.INT;
+};
+
+flexbuffers.ValueType.fixedTypedVectorElementType = (valueType) => {
+  return ((valueType - flexbuffers.ValueType.VECTOR_INT2) % 3) + flexbuffers.ValueType.INT;
+};
+
+flexbuffers.ValueType.fixedTypedVectorElementSize = (valueType) => {
+  // The x / y >> 0 trick is to have an int division. Suppose to be faster than Math.floor()
+  return (((valueType - flexbuffers.ValueType.VECTOR_INT2) / 3) >> 0)  + 2;
+};
+
+flexbuffers.ValueType.packedType = (valueType, bitWidth) => {
+  return bitWidth | (valueType << 2);
+};
+
+flexbuffers.toReference = (buffer) => {
+
+  function validateOffset(dataView, offset, width) {
+    if (dataView.byteLength <= offset + width || offset & (flexbuffers.BitWidth.toByteWidth(width) - 1) !== 0) {
+      throw "Bad offset: " + offset + ", width: " + width;
+    }
+  }
+
+  function readInt(dataView, offset, width) {
+    validateOffset(dataView, offset, width);
+    if (width === flexbuffers.BitWidth.WIDTH8) {
+      return dataView.getInt8(offset);
+    }
+    if (width === flexbuffers.BitWidth.WIDTH16) {
+      return dataView.getInt16(offset, true);
+    }
+    if (width === flexbuffers.BitWidth.WIDTH32) {
+      return dataView.getInt32(offset, true);
+    }
+    return dataView.getBigInt64(offset, true);
+  }
+
+  function readUInt(dataView, offset, width) {
```

Comment:
This function is the most frequently called of all, so its performance really matters. Much like the C++ code, you might get a tiny bit more speed by splitting the cases in two. Also worried about all those constant "lookups", might be worth using integers here :)

Reply:
Using integer now. 

> Much like the C++ code, you might get a tiny bit more speed by splitting the cases in two.

@aardappel could you please give me a link to the code, at the first glance I could not find it. Haven't checked C++ implementation for 3 years now :)

Reply:
https://github.com/google/flatbuffers/blob/e5a8f76a4b66f738b964b1af58df990229440f27/include/flatbuffers/flexbuffers.h#L130-L168

It be good to have a look thru the C++ code.. it is the most correct, feature-full, and close to the design intent version we have.

Reply:
I changed the function to reflect similar flow as in C++ and also introduced a check if BigInt/BigUint is supported on the platform. In case it is not it creates and object with low, high property, similar to how it works in FlatBuffers JS.